Human Life Optimization: Citrus Flavored

LemonInTheDark opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

### Help, what's going on

Me and kyler have made a blood pact to make `Life()` cheaper
He's slow, so here's what I've come up with.

Most of what I've focused on is eliminating unneeded work. I've done a few smaller optimizations, mostly because I was bored. They're not things to replicate I don't think.

Oh and uh, I'm gonna be giving percentages of "`Life()`" as measures for the impact of changes. 
These are not percentages of the total cost of `Life()`, but rather the total cost of the base human Life (It's hard to test say, digestion when you're on a local for instance). 
The upper bound on the impact any changes here would make is roughly 2/3rds their percentage. Just so you know  

I'ma go through change by change.

### Change by Change

`Life()` Optimization Part 1: Organ order (7ffd6fe7f5e0354ca01302057bc736f7214a06b3)
It turns out organ code is really quite dumb. Wastes a lot of time.
I'm going to be doing what I can to optimize it in this branch.

To start with:

If we can ensure organs are sorted according to `organ_process_order` inside `internal_organs_slot`
We can do away with a needless get_organ_slot call in `handle_organs`, and avoid needing to iterate over 30 entries, one per possible organ slot, instead iterating just the ones we have

This saves roughly 5% of life

Organ Optimization Part 2: base on_life()(ff610653390011f38d2474c8beb2e575193005f6)

Organ `on_life`, despite not looking it, is a remarkably hot proc. Called something close to 9 times for each human on the map, it's worth taking care to clean it up.

As things currently stand, we do two dumb things.
The first is manually checking for organ failure each iteration.
We do this because we allow organ damage to be modified outside of the `applyOrganDamage` proc, which also calls `check_failing_thresholds`.

There's no reason to do this, so I've gone through and removed all instances of it

The second is calling `applyOrganDamage` no matter what, to "heal" the organ. Even if it isn't damaged.

The fix for this is simple, just an if check.

This saves roughly 10% of pre changes `Life()` cost

Blood code is slightly more sane, but it calls `get_part_bleed_rate` a lot, and does other checks that are the same so long as the bleed rate never changes. 
This is reasonably expensive

So, I'm going to make it event driven, and cache the bleed rate. 
But to do that, I need to be able to react to limb ownership changes, and well,there's a few that don't use the existing setter. 
This commit (c94b625e59b387d472ee9d0a6abb69627a3b3d1c) fixes that 

Basically, anything that effects bleed rate needs to be setup so it can be reacted to as an event. 
Mostly by setters.
This commit (8f3921c7f309018d4349f5f2d258c2cc0d494d88) makes `NOBLEED` and `generic_bleedstacks` support that
(746db06a43570835ac66bf1fce439bc1c3bd141e) adds the same to embedded items.
(9650c476efe2fb1cdb8fb19507c53a82453e6966) Adds a setter for wound blood flow, and hooks it into `refresh_bleed_rate`

@Ryll-Ryll Yo, can you think of anything in wound code that would benefit from being able to react to blood flow changing?
I noticed some stuff that looked like it might, but I don't trust myself with stuff I'm not that comfortable with


(ca7f52ad0a929df2ddf12690637925c25cbdf44e) fully caches `bleed_rate`. Because I can do this, we only need to call `update_part_wound_overlay` when bleed rate updates
This saves 15%! of human life tick. Get owned

(8f4bb7f2284657398fdd37c01e7a09697a16ee4f) touches some temperature code.

It turns out that clearing alerts and sending signals every process is wasteful. 
There's no reason to do it unless last process was a problem one

The change for that is quite easy, literally just a new var on human.

Saves 2% of human life tick.
Very nice for the amount of time I put into it

BTW, I have a feeling that most of the overhead of bodytemp is caused by human body temperature being higher then room temperature. 
Not sure what to do about this though

(a8f04d293178e6f11954b364d7afe89aedc2ef2a) does something similar to disgust

Basically, don't continuously send signals if you have no disgust
It's rare enough that this is all the caching we really need to do

Saves roughly 1% of life tick. Right on the edge of not worth it

Oh and I cleaned up some of liver, gravity and fire lighting code. Nothing worth writing home about, but it makes things a bit nicer.


### Results 

That brings us to roughly 33% of base human `Life()` saved. So maybe 22% of highpop cost? 
Hard to say for sure. Regardless, I'm quite happy with this

If you're curious, here are all my profiles in order.
[Initial, just humans.txt](https://github.com/tgstation/tgstation/files/8440914/Initial.just.humans.txt)
[Post handle_organ and base on_life.txt](https://github.com/tgstation/tgstation/files/8440922/Post.handle_organ.and.base.on_life.optim.txt)
[Post organs and blood caching.txt](https://github.com/tgstation/tgstation/files/8440916/Post.organs.and.blood.caching.txt)
[Post alert cache.txt](https://github.com/tgstation/tgstation/files/8440915/Post.alert.cache.txt)
[Post stomach.txt](https://github.com/tgstation/tgstation/files/8440917/Post.stomach.txt)
[post gravity.txt](https://github.com/tgstation/tgstation/files/8440921/post.gravity.txt)


Oh. And uh, I found what looked like a bug in soundbanging logic.

If you get soundbangd and see the "You can't hear anything!" message, you will now properly go deaf untill you get healing
Rather then the current behavior of just... doing nothing.

It's technically a balance change, maybe I should atomize it? Thoughts?

## Why It's Good For The Game

[![IMAGE ALT TEXT](http://img.youtube.com/vi/atuFSv2bLa8/0.jpg)](http://www.youtube.com/watch?v=atuFSv2bLa8)

## Changelog

<!-- If your PR modifies aspects of the game that can be concretely observed by players or admins you should add a changelog. If your change does NOT meet this description, remove this section. Be sure to properly mark your PRs to prevent unnecessary GBP loss. You can read up on GBP and it's effects on PRs in the tgstation guides for contributors. Please note that maintainers freely reserve the right to remove and add tags should they deem it appropriate. You can attempt to finagle the system all you want, but it's best to shoot for clear communication right off the bat. -->

:cl:
balance: Getting your eardrums blown out will now properly have a low chance to make you go fully deaf, until you find some earmuffs or get a transplant
refactor: Cached some redundant operations in Life(), speeding it up by ~22%. I know that doesn't mean much to you, but cpu time go brrr ok? 
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

LemonInTheDark committed:
Life() Optimization Part 1: Organ order

It turns out organ code is really quite dumb. Wastes a lot of time.
I'm going to be doing what I can to optimize it in this branch.

To start with:

If we can ensure organs are sorted according to organ_process_order inside
internal_organs_slot
We can do away with a needless get_organ_slot call in handle_organs, and
avoid needing to iterate over 30 entries, one per possible organ slot,
instead iterating just the ones we have

This saves roughly 5% of life

LemonInTheDark committed:
Organ Optimization Part 2: base on_life()

Organ on_life, despite not looking it, is a remarkably hot proc. Called
something close to 9 times for each human on the map, it's worth taking
care to clean it up.

As things currently stand, we do two dumb things.
The first is manually checking for organ failure each iteration.
We do this because we allow organ damage to be modified outside of the
applyOrganDamage proc, which also calls check_failing_thresholds.

There's no reason to do this, so I've gone through and removed all
instances of it

The second is calling applyOrganDamage no matter what, to "heal" the
organ. Even if it isn't damaged.

The fix for this is simple, just an if check.

This saves roughly 10% of pre changes Life() cost

LemonInTheDark committed:
Blood cleanup beta version

Blood code is slightly more sane, but it calls get_part_bleed_rate a
lot, and does other checks that are the same so long as the bleed rate
never changes. This is reasonably expensive

So, I'm going to make it event driven, and cache the bleed rate. But to
do that, I need to be able to react to limb ownership changes, and well,
there's a few that don't use the existing setter. This commit fixes that

LemonInTheDark committed:
Blood cleanup beta 2.0

I'm starting on the work to make bleed rate cached

Adds NO_BLEED to the signal registration of set_owner
This lets us generically react to it

Makes changing your bleedstacks into a proc we can hook into
Fully implements it

LemonInTheDark committed:
Blood optimization beta 3.0

Hooks into embedding/embed details updating, this ensures the embed
aspect of bleed rate is properly accounted for

LemonInTheDark committed:
Blood optimization beta 4.0

Hooks wounds into refresh_bleed_rate, adds a setter for blood flow to
support this

I feel like there are places where this would be useful, but I'm not
totally sure
Kinda wanted to ask ryll

LemonInTheDark committed:
Blood optimization, the final countdown

Fully caches bleed_rate. Because I can do this, we only need to call update_part_wound_overlay when bleed rate updates
This saves 15%! of human life tick. Get owned

LemonInTheDark committed:
Optimizes body_temperature_alerts

It turns out that clearing alerts and sending signals every process is
wasteful. There's no reason to do it unless last process was a problem
one

The change for that is quite easy, literally just a new var on human.

Saves 2% of human life tick.
Very nice for the amount of time I put into it

BTW, I have a feeling that most of the overhead of bodytemp is caused by
human body temperature being higher then room temperature. Not sure
what to do about this though

LemonInTheDark committed:
Does something similar to disgust

Basically, don't continusly send signals if you have no disgust
It's rare enough that this is all the caching we really need to do

Saves roughly 1% of life tick. Right on the edge of not worth it

LemonInTheDark committed:
Some cleanup to how mob fires handle their lighting, adds a check to liver coder

LemonInTheDark committed:
Cleans up update_gravity slightly

It's very close to outside a measurable deviation faster, but I honestly
just did it to make it easier to read

LemonInTheDark committed:
Fuck you

Kapu1178 commented on code/modules/surgery/bodyparts/_bodyparts.dm:
```diff
@@ -784,42 +798,94 @@
+		|| old_bleedstacks > 0 && generic_bleedstacks <= 0)
+		refresh_bleed_rate()
+
+/obj/item/bodypart/proc/on_owner_nobleed_loss(datum/source)
+	SIGNAL_HANDLER
+	refresh_bleed_rate()
+
+/obj/item/bodypart/proc/on_owner_nobleed_gain(datum/source)
+	SIGNAL_HANDLER
+	refresh_bleed_rate()
```
```suggestion
/obj/item/bodypart/proc/on_owner_nobleed_loss(datum/source)
	SHOULD_CALL_PARENT(TRUE)
	PROTECTED_PROC(TRUE)
	SIGNAL_HANDLER
	refresh_bleed_rate()

/obj/item/bodypart/proc/on_owner_nobleed_gain(datum/source)
	SHOULD_CALL_PARENT(TRUE)
	PROTECTED_PROC(TRUE)
	SIGNAL_HANDLER
	refresh_bleed_rate()
```

LemonInTheDark replied:
I uh, why? is it just to mark out to people that they shouldn't be calling these things manually? I didn't think that happen that often.

I'm now noticing that it's consistent across the file, so I'm ok with it on that merit, but it seems a bit strange to me.

Kapu1178 replied:
Yes. Linters are significantly more effective than comments.

LemonInTheDark replied:
I'm not a fan of this pattern, it seems weird to need to lint against people calling signal procs directly, seems like fluff? idk

Kapu1178 commented on code/modules/surgery/bodyparts/_bodyparts.dm:
```diff
@@ -784,42 +798,94 @@
+	refresh_bleed_rate()
+
+/// Sets our generic bleedstacks
+/obj/item/bodypart/proc/setBleedStacks(set_to)
+	adjustBleedStacks(set_to - generic_bleedstacks)
```
```suggestion
/obj/item/bodypart/proc/setBleedStacks(set_to)
	SHOULD_CALL_PARENT(TRUE)
	adjustBleedStacks(set_to - generic_bleedstacks)
```

LemonInTheDark replied:
Why? I'm perfectly fine with someone overriding this if they really wanted to. Is it because they should instead be overriding adjustBleedStacks? Is that worth the extra line?

Kapu1178 replied:
Yes.

LemonInTheDark replied:
I'm not sure I agree, but alright

LemonInTheDark replied:
actually I just like this, good point

Kapu1178 commented on code/modules/surgery/bodyparts/_bodyparts.dm:
```diff
@@ -784,42 +798,94 @@
+	embedded_objects -= unembed
+	refresh_bleed_rate()
+
+/obj/item/bodypart/proc/embedded_object_changed(obj/item/embedded_source)
+	SIGNAL_HANDLER
```
```suggestion
/obj/item/bodypart/proc/embedded_object_changed(obj/item/embedded_source)
	PROTECTED_PROC(TRUE)
	SIGNAL_HANDLER
```

LemonInTheDark replied:
Why no should call parent here? what's the difference?

Kapu1178 replied:
PROTECTED_PROC means the proc cannot be called outside of bodypart code, mostly to prevent idiocy.

LemonInTheDark replied:
do you think that's worth the extra line? I guess you would, since you made it.
as a general pattern I wouldn't like it, too much fluff

LemonInTheDark replied:
see above one

Kapu1178 reviewed (changes requested):
Spaceman DMM flags on bodypart procs pls.

Also, why did codeowners bot not ping me?

ATH1909 commented:
absolute legend

LemonInTheDark commented:
> Spaceman DMM flags on bodypart procs pls.

I'm honestly a bit confused about the convention you're operating under. I'm ok with matching the file cause that's just good manners, but some of your suggestions seem a bit strange.
 
> Also, why did codeowners bot not ping me?

It does that sometimes. I think it's bugged? there are some members who are just never alerted, or are alerted later. It's weird.

LemonInTheDark committed:
Very minor organ optimization

Instead of using internal_organs_slot for our organ iteration,
we use it to produce a sorted internal_organs list.

This is barely worth it, ends up being 0.7% of life saved.
I only did it because it makes a semi noticable impact on our current
numbers.

LemonInTheDark committed:
Revert "Very minor organ optimization"

It turns out mutating internal_organs is important sometimes
The usefulness of this change is small enough that I don't want to spend
more time on it

So back it goes into the dirt
This reverts commit 622bd34adb8d8a3d5f2763ac659446ef3362e3cd.

AnturK commented on code/modules/mob/living/living.dm:
```diff
@@ -1394,16 +1399,16 @@
-		return TRUE
-	return FALSE
+	if(fire_stacks <= 0 || on_fire)
+		return
```
This should probably return FALSE

LemonInTheDark replied:
oh yeah good point

AnturK reviewed (commented):
Was this testmerged ?

LemonInTheDark commented:
> Was this testmerged ?

It wasn't, good point

LemonInTheDark commented on code/modules/mob/living/living.dm:
```diff
@@ -1394,16 +1399,16 @@
-		return TRUE
-	return FALSE
+	if(fire_stacks <= 0 || on_fire)
+		return
```
```suggestion
		return FALSE
```

LemonInTheDark committed:
Properly returns false in IgniteMob

Kapu1178 commented on code/modules/surgery/bodyparts/_bodyparts.dm:
```diff
@@ -784,42 +798,94 @@
- * * ignore_modifiers - If TRUE, ignore the bleed reduction for laying down and grabbing your limb
- */
-/obj/item/bodypart/proc/get_part_bleed_rate(ignore_modifiers = FALSE)
+/// INTERNAL PROC, DO NOT USE
+/// Properly sets us up to manage an inserted embeded object
+/obj/item/bodypart/proc/_embed_object(obj/item/embed)
+	if(embed in embedded_objects) // go away
```
```suggestion
/// INTERNAL PROC, DO NOT USE
/// Properly sets us up to manage an inserted embeded object
/obj/item/bodypart/proc/_embed_object(obj/item/embed)
	PROTECTED_PROC(TRUE)
	if(embed in embedded_objects) // go away
```
if you dont want it used outside of internals, use the SMDMM lint for it ya dingus

LemonInTheDark replied:
Good point

LemonInTheDark replied:
actually, I don't think I can, since it's not called by a bodypart

Kapu1178 commented on code/modules/surgery/bodyparts/_bodyparts.dm:
```diff
@@ -784,42 +798,94 @@
+
+/// INTERNAL PROC, DO NOT USE
+/// Cleans up any attachment we have to the embedded object, removes it from our list
+/obj/item/bodypart/proc/_unembed_object(obj/item/unembed)
+	UnregisterSignal(unembed, COMSIG_ITEM_EMBEDDING_UPDATE)
```
```suggestion
/obj/item/bodypart/proc/_unembed_object(obj/item/unembed)
	PROTECTED_PROC(TRUE)
	UnregisterSignal(unembed, COMSIG_ITEM_EMBEDDING_UPDATE)
```
same here

LemonInTheDark replied:
yeet

LemonInTheDark replied:
unfortunately ditto

LemonInTheDark committed:
Adds a parent call to setBleedStacks

I like it, it's a good idea

LemonInTheDark committed:
Merge branch 'the-void-stares-back' of https://github.com/LemonInTheDark/tgstation into the-void-stares-back
